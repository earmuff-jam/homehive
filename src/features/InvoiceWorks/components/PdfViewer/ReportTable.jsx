import React from "react";

import {
  Box,
  Paper,
  Stack,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
} from "@mui/material";

export default function ReportTable({
  rows = [],
  taxRate = 0,
  showWatermark,
  invoiceStatus,
  invoiceTitle = "Invoice Details",
}) {
  const numberFormatter = (number) => Number(number).toFixed(2);

  const subtotal = rows
    .map(({ price }) => parseFloat(price) || 0)
    .reduce((sum, i) => sum + i, 0);

  const paymentRecieved = rows
    .map(({ payment }) => parseFloat(payment) || 0)
    .reduce((sum, i) => sum + i, 0);

  const invoiceSubtotal = subtotal - paymentRecieved;
  const formattedTax = parseFloat(taxRate) || 0;
  const invoiceTaxes = (formattedTax / 100) * invoiceSubtotal;
  const invoiceTotal = invoiceSubtotal + invoiceTaxes;

  return (
    <TableContainer component={Paper}>
      <Table sx={{ minWidth: 700 }} aria-label="spanning table">
        <TableHead>
          <TableRow>
            <TableCell align="center" colSpan={5} sx={{ fontWeight: "bold" }}>
              {invoiceTitle}
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              Price (USD)
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell sx={{ fontWeight: "bold" }}>Category</TableCell>
            <TableCell sx={{ fontWeight: "bold" }}>Description</TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              Qty.
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              Cost
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              Payment Received
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              Balance Due
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {rows.map((row, index) => (
            <TableRow key={index}>
              <TableCell>
                <Typography variant="subtitle">
                  {row?.category || ""}
                </Typography>
              </TableCell>
              <TableCell>
                <Stack direction="row" spacing={1} alignItems="flex-end">
                  <Typography variant="subtitle">
                    {row.descpription || ""}
                  </Typography>
                  <Typography variant="caption">
                    <sub>{row.caption || ""}</sub>
                  </Typography>
                </Stack>
              </TableCell>
              <TableCell align="right">{row.quantity}</TableCell>
              <TableCell align="right">{numberFormatter(row.price)}</TableCell>
              <TableCell align="right">
                {numberFormatter(row.payment)}
              </TableCell>
              <TableCell align="right">
                {numberFormatter(
                  (parseFloat(row.price) || 0) - (parseFloat(row.payment) || 0),
                )}
              </TableCell>
            </TableRow>
          ))}
          <TableRow>
            <TableCell rowSpan={3} />
            <TableCell colSpan={4} sx={{ fontWeight: "bold" }}>
              Subtotal
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              {numberFormatter(invoiceSubtotal)}
            </TableCell>
          </TableRow>

          <TableRow>
            <TableCell sx={{ fontWeight: "bold" }}>Tax</TableCell>
            <TableCell colSpan={3} align="right" sx={{ fontWeight: "bold" }}>
              {formattedTax}%
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              {numberFormatter(invoiceTaxes)}
            </TableCell>
          </TableRow>

          <TableRow>
            <TableCell colSpan={4} sx={{ fontWeight: "bold" }}>
              <Stack direction="row" spacing={10}>
                <Typography>Total</Typography>
                <Box component="span">
                  {invoiceStatus?.display && (
                    <Typography
                      color="error.light"
                      variant="caption"
                      className={!showWatermark && "no-print"} // print only when allowed
                      sx={{
                        textTransform: "uppercase",
                        fontSize: "2rem",
                        padding: "0.3rem 0.5rem",
                        textAlign: "center",
                        border: "0.3rem solid",
                      }}
                    >
                      {invoiceStatus?.label}
                    </Typography>
                  )}
                </Box>
              </Stack>
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              {numberFormatter(invoiceTotal)}
            </TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </TableContainer>
  );
}
