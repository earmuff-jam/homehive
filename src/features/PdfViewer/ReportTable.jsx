import {
  Paper,
  Table,
  TableCell,
  TableContainer,
  TableHead,
  TableBody,
  TableRow,
  Typography,
  Stack,
} from "@mui/material";

export default function ReportTable({
  rows = [],
  taxRate = 0,
  invoiceTitle = "Rent Details",
}) {
  const numberFormatter = (number) => number.toFixed(2);
  const subtotal = rows
    .map(({ price }) => parseFloat(price))
    .reduce((sum, i) => sum + i, 0);

  const paymentRecieved = rows
    .map(({ payment }) => parseFloat(payment) || 0)
    .reduce((sum, i) => sum + i, 0);

  const invoiceSubtotal = subtotal - paymentRecieved;
  const formattedTax = parseFloat(taxRate) || 0;
  const invoiceTaxes = formattedTax * invoiceSubtotal;
  const invoiceTotal = invoiceTaxes + invoiceSubtotal;

  return (
    <TableContainer component={Paper}>
      <Table sx={{ minWidth: 700 }} aria-label="spanning table">
        <TableHead>
          <TableRow>
            <TableCell align="center" colSpan={4}>
              {invoiceTitle}
            </TableCell>
            <TableCell align="right">Price (USD) </TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Description</TableCell>
            <TableCell align="right">Qty.</TableCell>
            <TableCell align="right">Cost</TableCell>
            <TableCell align="right">Payment Recieved</TableCell>
            <TableCell align="right">Balance Due</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {rows.map((row) => (
            <TableRow key={row.id}>
              <TableCell>
                <Stack direction="row" spacing={1} alignItems="flex-end">
                  <Typography variant="subtitle">{row.descpription}</Typography>
                  <Typography variant="caption">{row.caption}</Typography>
                </Stack>
              </TableCell>
              <TableCell align="right">{row.quantity}</TableCell>
              <TableCell align="right">{row.price}</TableCell>
              <TableCell align="right">{row.payment}</TableCell>
              <TableCell align="right">
                {numberFormatter(row.price - row.payment)}
              </TableCell>
            </TableRow>
          ))}
          <TableRow>
            <TableCell rowSpan={3} />
            <TableCell colSpan={3}>Subtotal</TableCell>
            <TableCell align="right">
              {numberFormatter(invoiceSubtotal)}
            </TableCell>
          </TableRow>

          <TableRow>
            <TableCell>Tax</TableCell>
            <TableCell
              colSpan={1}
              align="right"
            >{`${formattedTax} %`}</TableCell>
            <TableCell align="right" colSpan={2}>
              {numberFormatter(formattedTax)}
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell colSpan={3}>Total</TableCell>
            <TableCell align="right">{numberFormatter(invoiceTotal)}</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </TableContainer>
  );
}
