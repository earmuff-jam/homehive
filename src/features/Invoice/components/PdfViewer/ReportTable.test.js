// ReportTable.test.jsx
import React from "react";

import { MemoryRouter } from "react-router-dom";

import ReportTable from "./ReportTable";
import { render, screen } from "@testing-library/react";

describe("Report Table Component tests", () => {
  describe("Report Table Snapshot tests", () => {
    test("PdfViewer snapshot", () => {
      const { asFragment } = render(
        <MemoryRouter>
          <ReportTable />
        </MemoryRouter>,
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe("Report Table tests", () => {
    describe("ReportTable", () => {
      const mockRows = [
        {
          category: "Rent",
          descpription: "Monthly rent",
          caption: "April",
          quantity: 1,
          price: 1000,
          payment: 500,
        },
      ];

      const mockInvoiceStatus = {
        label: "Paid",
        display: true,
      };

      it("renders without crashing and displays data", () => {
        render(
          <ReportTable
            rows={mockRows}
            taxRate={10}
            showWatermark={false}
            invoiceStatus={mockInvoiceStatus}
            invoiceTitle="Test Invoice"
          />,
        );

        expect(screen.getByText(/Test Invoice/i)).toBeInTheDocument();

        expect(screen.getByText(/Monthly rent/i)).toBeInTheDocument();
        expect(screen.getByText(/April/i)).toBeInTheDocument();
      });
    });
  });
});
