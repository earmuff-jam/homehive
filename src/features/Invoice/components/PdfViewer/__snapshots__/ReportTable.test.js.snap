// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report Table Component tests Report Table Snapshot tests PdfViewer snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <table
      aria-label="spanning table"
      class="MuiTable-root css-vf0yke-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1xtozoh-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-774dy1-MuiTableCell-root"
            colspan="5"
            scope="col"
          >
            Invoice Details
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1sipjqd-MuiTableCell-root"
            scope="col"
          >
            Price (USD)
          </th>
        </tr>
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1xtozoh-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-155k23g-MuiTableCell-root"
            scope="col"
          >
            Category
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-155k23g-MuiTableCell-root"
            scope="col"
          >
            Description
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1sipjqd-MuiTableCell-root"
            scope="col"
          >
            Qty.
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1sipjqd-MuiTableCell-root"
            scope="col"
          >
            Cost
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1sipjqd-MuiTableCell-root"
            scope="col"
          >
            Payment Received
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1sipjqd-MuiTableCell-root"
            scope="col"
          >
            Balance Due
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            rowspan="3"
          />
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-qq3g53-MuiTableCell-root"
            colspan="4"
          >
            Subtotal
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1ri9by7-MuiTableCell-root"
          >
            0.00
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-qq3g53-MuiTableCell-root"
          >
            Tax
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1ri9by7-MuiTableCell-root"
            colspan="3"
          >
            0%
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1ri9by7-MuiTableCell-root"
          >
            0.00
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-qq3g53-MuiTableCell-root"
            colspan="4"
          >
            <div
              class="MuiStack-root css-7v40v5-MuiStack-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
              >
                Total
              </p>
              <span
                class="MuiBox-root css-0"
              />
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1ri9by7-MuiTableCell-root"
          >
            0.00
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
